Refactor pypeit

profxj opened the pull request:
As titled, this PR is a major refactor of the primary module of PypeIt.

The main goals are to:

- Expose the primary methods currently in `pypeit.py` which perform image processing, object finding, and extraction
- Enable users to perform "step by step" reductions on individual detectors/mosaics.
- Clean up some of the chaff that has grown in `pypeit.py` over the years
- Enable spectrographs to effectively skip the processing step, e.g. if they have their own tried+true processing codes
- Put in the hooks for the Dashboard that will allow step by step running and monitoring of PypeIt

The new script is `pypeit_reduce_by_step` which runs on a single science/standard frame (or a set of frames, if `comb_id` is set in the PypeIt file) and a single detector/mosaic of that frame. 

Docs are provided with warnings about running PypeIt in anything but the end-to-end mode

I have been testing in the [dev_pypeit_refactor](https://github.com/pypeit/PypeIt-development-suite/tree/dev_pypeit_refactor) branch in the DevSuite.  Thus far, these are ad hoc tests and I'll be
begging @badpandabear 's help to meld them into standard DevSuite runs.  Those tests do check
that we are achieving nearly identical results with the new vs. old PypeIt as well as with the new
`pypeit_reduce_by_step` script.

I should add that the `state.py` code should be ignored as well as any hooks to it,
unless you see a way things could break.  For now, there are hooks but none of that
code should be active.

profxj committed:
wip

profxj committed:
slits

profxj committed:
wip

profxj committed:
Merge remote-tracking branch 'origin/state_and_steps' into deep_state

profxj committed:
Merge remote-tracking branch 'origin/state_and_steps' into deep_state

profxj committed:
progress

profxj committed:
Merge remote-tracking branch 'origin/state_and_steps' into deep_state

profxj committed:
working example

profxj committed:
Merge remote-tracking branch 'origin/develop' into deep_state

profxj committed:
load, etc.

profxj committed:
a few advances

profxj committed:
tilts

profxj committed:
flats

profxj committed:
2 detectors tested

profxj committed:
det, calibID

profxj committed:
fussin and fussin

profxj committed:
closer now

profxj committed:
yet mo

profxj committed:
extract + save images

profxj committed:
full run

profxj committed:
add sign

profxj committed:
oh, the near-IR

profxj committed:
mo

profxj committed:
got it?

profxj committed:
calib_ID

profxj committed:
y long

profxj committed:
Merge remote-tracking branch 'origin/develop' into refactor_pypeit

profxj committed:
mo

profxj committed:
mo

profxj committed:
debuggin

profxj committed:
deep in debug

profxj committed:
deeper

profxj committed:
still at it

profxj committed:
success

profxj committed:
reduce_exposure

profxj committed:
save_exposure

profxj committed:
mo wip

profxj committed:
refactoring continues

profxj committed:
broken up

profxj committed:
step by step

profxj committed:
new script

profxj committed:
uh oh..

profxj committed:
standard

profxj committed:
exposure docs

profxj committed:
docs galore

profxj committed:
outputfiles

profxj committed:
ready to test

profxj committed:
debuggin

profxj committed:
??

profxj committed:
mo

profxj committed:
mo

profxj committed:
cailb_all

profxj committed:
scrub the state for now

profxj committed:
Merge remote-tracking branch 'origin/release' into refactor_pypeit

profxj committed:
Merge remote-tracking branch 'origin/develop' into refactor_pypeit

profxj committed:
fussin

profxj committed:
closer now

profxj committed:
mo

profxj committed:
mo

profxj committed:
more docs

profxj committed:
Merge remote-tracking branch 'origin/develop' into refactor_pypeit

profxj committed:
docs

profxj committed:
refactor on top of pypeit

profxj committed:
path

profxj committed:
mo docs

profxj committed:
datetime fix

profxj committed:
test fix

profxj committed:
disabling state

kbwestfall commented on doc/reduce_by_step.rst:
```diff
@@ -0,0 +1,237 @@
+Optional Arguments
+~~~~~~~~~~~~~~~~~~
+
+- ``--show``: Show reduction steps via plots and outputs to ginga (requires remote control ginga session)
```
I added the `-h` output, like we have for other scripts, but I left the other text as is.  I'd prefer that we rely on the `-h` as much as possible so that we only have to maintain the help dialog in one place.

profxj replied:
good call!

kbwestfall commented on doc/reduce_by_step.rst:
```diff
@@ -0,0 +1,237 @@
+- Loads objects and sky models from the findobj step
+- Performs optimal extraction of 1D spectra
+- Applies wavelength calibration
+- Handles flux calibration (if standard stars available)
```
What do you mean by this?  Does the extraction now also do the flux calibration?

profxj replied:
whoops, chatbot hallucination 
  fixed

kbwestfall committed:
Merge branch 'develop' into refactor_pypeit

kbwestfall committed:
doc update

kbwestfall committed:
doc update

kbwestfall committed:
add reduce_by_step to script imports and update docs

kbwestfall commented on pypeit/core/findobj_skymask.py:
```diff
@@ -1414,6 +1415,9 @@ def objfind_QA(spat_peaks, snr_peaks, spat_vector, snr_vector, snr_thresh, qa_ti
     if objfindQA_filename is not None:
+        # Genreate the output folders
+        if not os.path.isdir(os.path.dirname(objfindQA_filename)):
+            os.makedirs(os.path.dirname(objfindQA_filename))
```
Use `Path` instead:
```
if objfindQA_filename is not None:
    qafile = Path(objfindQA_filename).absolute()
    qafile.parent.mkdir(parents=True, exists_ok=True)
    fig.savefig(qafile, dpi=400)
```

kbwestfall commented on pypeit/images/pypeitimage.py:
```diff
@@ -549,8 +549,10 @@ def build_mask(self, saturation=None, mincounts=None, slitmask=None, from_scratc
+            bpm = self.fullmask.flagged('BPM')
+            cr = self.fullmask.flagged('CR')
+            #bpm = self.fullmask.bpm
+            #cr = self.fullmask.cr
```
rm comments

profxj replied:
gone

kbwestfall commented on pypeit/scripts/ql.py:
```diff
@@ -779,7 +783,7 @@ def main(args):
         # Find the raw science files
         sci_idx = ps.fitstbl.find_frames('science') if args.sci_files is None \
-                        else np.in1d(ps.fitstbl['filename'].data, args.sci_files)
+                        else np.isin(ps.fitstbl['filename'].data, args.sci_files)
```
Good call.  There were only a few instances of this, so I converted all the `in1d` instances into `isin`.

profxj replied:
thanks!

kbwestfall committed:
in1d -> isin

kbwestfall commented on pypeit/scripts/ql.py:
```diff
@@ -299,9 +300,12 @@ def generate_sci_pypeitfile(redux_path:str,
-                            ps_sci.fitstbl.construct_basename(i))
+            #std_spec1d = pypeit.PypeIt.get_spec_file_name(
+            #                sci_dir / ps_sci.par['rdx']['scidir'],
+            #                ps_sci.fitstbl.construct_basename(i))
```
rm comments

profxj replied:
done

kbwestfall commented on pypeit/scripts/reduce_by_step.py:
```diff
@@ -0,0 +1,279 @@
+        pypeIt.reuse_calibs = True
+
+        # Detector
+        # TODO -- worry about mosaics
```
Is this still a TODO?

profxj replied:
not so much;  removed

kbwestfall commented on pypeit/scripts/reduce_by_step.py:
```diff
@@ -0,0 +1,279 @@
+            print(f"No detector provided (--det). Choose from one of these: {detectors}.") 
+            if len(mosaics) > 0:
+                print("")
+                print(f"This instrument also supports mosaics: {mosaics}")
```
"... also supports *the following* mosaics: ..."

profxj replied:
added

kbwestfall commented on pypeit/scripts/reduce_by_step.py:
```diff
@@ -0,0 +1,279 @@
+            if det not in mosaics:
+                msgs.error(f"Mosaic {args.det} not found. Choose from one of these: {mosaics}.")
+        else:
+            msgs.error(f"Detector {args.det} not found. Choose from one of these: {detectors}.")
```
For all of this parsing whether or not the provided `det` values are right, can we instead use the `select_detectors` function?   We can move some of the useful error messages provided here into that function, right?

profxj replied:
attempted!

kbwestfall commented on pypeit/scripts/reduce_by_step.py:
```diff
@@ -0,0 +1,279 @@
+        row = np.where(pypeIt.fitstbl['filename'] == args.frame)[0]
+        if len(row) != 1:
+            msgs.error(f"Frame {args.frame} not found or not unique")
+        frame = int(row)
```
I think this throws a deprecation warning.  I would do `frame = row[0]` instead, like you do below.

profxj replied:
modified

kbwestfall commented on pypeit/scripts/reduce_by_step.py:
```diff
@@ -0,0 +1,279 @@
+        frames = np.where(pypeIt.fitstbl['comb_id'] == comb_id)[0]
+        # Find all frames whose comb_id matches the current frames bkg_id.
+        bg_frames = np.where((pypeIt.fitstbl['comb_id'] == pypeIt.fitstbl['bkg_id'][frames][0])
+                                & (pypeIt.fitstbl['comb_id'] >= 0))[0]
```
This should be a method in `fitstbl`.

profxj replied:
good suggestion;  done

kbwestfall commented on pypeit/scripts/reduce_by_step.py:
```diff
@@ -0,0 +1,279 @@
+
+        # Intermediate filenames
+        sci_filename = outputfiles.intermediate_filename('sciImg', basename, 
+                                    pypeIt.spectrograph.get_det_name(det))
```
Do `pypeIt.spectrograph.get_det_name(det)` once.

profxj replied:
agreed

kbwestfall commented on pypeit/scripts/reduce_by_step.py:
```diff
@@ -0,0 +1,279 @@
+            if bkg_redux_sciimg is not None:
+                bkg_redux_sciimg.to_file(bkg_filename, overwrite=True)
+                msgs.info(f'Wrote intermediate background image to {bkg_filename}')
+            '''
```
remove?

kbwestfall commented on pypeit/scripts/reduce_by_step.py:
```diff
@@ -0,0 +1,279 @@
+
+            # All done
+            return
+        else: # Load
```
Don't need the else.

profxj replied:
good call

kbwestfall commented on pypeit/scripts/reduce_by_step.py:
```diff
@@ -0,0 +1,279 @@
+                    std_outfile = outputfiles.get_std_outfile(pypeIt.fitstbl, pypeIt.par, 
+                                                          frame_indx[is_standard])
+                except pypmsgs.PypeItError:
+                    msgs.warn('No standard star spec1d file found for this science frame, but one was expected.') 
```
Why is it expected?

profxj replied:
because it is given in the PypeIt file

profxj replied:
comment expanded

kbwestfall commented on pypeit/scripts/reduce_by_step.py:
```diff
@@ -0,0 +1,279 @@
+                                                          frame_indx[is_standard])
+                except pypmsgs.PypeItError:
+                    msgs.warn('No standard star spec1d file found for this science frame, but one was expected.') 
+                    msgs.warn('Continuing without standard star information.')
```
Combine the warning statements.

profxj replied:
ok

kbwestfall commented on pypeit/scripts/reduce_by_step.py:
```diff
@@ -0,0 +1,279 @@
+                find_negative=find_negative,
+                std_outfile=std_outfile, show=args.show)
+
+            # TODO -- write from findobj_on_det?
```
Still a TODO?

profxj replied:
I can't remember.. :)

kbwestfall commented on pypeit/scripts/reduce_by_step.py:
```diff
@@ -0,0 +1,279 @@
+
+            # All done
+            return
+        else: # Load
```
Don't need the else.

profxj replied:
agreed!

kbwestfall commented on pypeit/scripts/reduce_by_step.py:
```diff
@@ -0,0 +1,279 @@
+            #
+            specobjs_objfind = specobjs.SpecObjs.from_fitsfile(spec1d_filename)
+
+        # TODO -- Add objs in here!! -- what did I mean by this?
```
I like the conversations with oneself :)  Still a TODO?

profxj replied:
not sure :)

kbwestfall commented on pypeit/scripts/reduce_by_step.py:
```diff
@@ -0,0 +1,279 @@
+            if pypeIt.par['reduce']['slitmask']['assign_obj']:
+                sciImg_dict = {}
+                calib_slits = []
+                # TODO - ask Debora to check this..
```
Flagging for @debora-pe .  To me, this seems like something that should be done as part of the `findobj` step.

kbwestfall commented on pypeit/scripts/reduce_by_step.py:
```diff
@@ -0,0 +1,279 @@
+                                   pypeIt.par, frame, all_spec2d, sobjs_extract,
+                                   pypeIt.calibrations_path,
+                                   in_update_det=det)
+        return 0
```
Don't need the return.

profxj replied:
removed

kbwestfall commented on pypeit/scripts/reduce_by_step.py:
```diff
@@ -0,0 +1,279 @@
+                                   pypeIt.calibrations_path,
+                                   in_update_det=det)
+        return 0
+
```
I think we're doing too much in the script.  Ideally, our scripts would be simple wrappers of short code snippets, and all the underlying complexity would be performed outside of the script.  E.g., the `run_pypeit.py` script is pretty simple because all the complexity is shuffled off elsewhere.  Admittedly, `run_pypeit` is a rare example of when we achieve this, so make of this comment what you will.

kbwestfall commented on pypeit/scripts/run_to_calibstep.py:
```diff
@@ -80,7 +80,7 @@ def main(args):
             row = np.where(pypeIt.fitstbl['filename'] == args.science_frame)[0]
             if len(row) != 1:
-                msgs.error(f"Frame {args.frame} not found or not unique")
+                msgs.error(f"Frame {args.science_frame} not found or not unique")
```
Good catch.  We need a test!

kbwestfall commented on pypeit/scripts/view_fits.py:
```diff
@@ -29,6 +29,8 @@ def get_parser(cls, width=None):
                             '--proc must be set in order for this option to work.')
 
+        parser.add_argument('--inter', default=False, action='store_true',
+                            help='Input file is an Intermediate SciImage file')
```
Can we detect this automatically?

profxj replied:
the current code is too painful to do so

kbwestfall commented on pypeit/tests/test_calibrations.py:
```diff
@@ -48,43 +51,52 @@ def multi_caliBrate(fitstbl):
+#    # Set
+#    det = 1
+#    calib.set_config(frame, det)
+#    return calib
```
remove?

profxj replied:
gone

kbwestfall commented on pypeit/calibrations.py:
```diff
@@ -303,37 +312,37 @@ def find_calibrations(self, frametype, frameclass):
+#        self.calib_ID = self.fitstbl.find_frame_calib_groups(self.frame)[0]
+#        self.det = det
+#        if par is not None:
+#            self.par = par
```
remove?

profxj replied:
gone

kbwestfall commented on pypeit/calibrations.py:
```diff
@@ -365,6 +374,7 @@ def get_arc(self, force:str=None):
 
         # If a processed calibration frame exists and we want to reuse it, do
         # so:
+        #embed(header='385 of calibrations.py')
```
remove?

profxj replied:
removed

kbwestfall commented on pypeit/calibrations.py:
```diff
@@ -525,11 +546,11 @@ def get_bias(self, force:str=None):
         # Find the calibrations
         frame = {'type': 'bias', 'class': buildimage.BiasImage}
-        raw_files, cal_file, calib_key, setup, calib_id, detname \
+        self.raw_files, cal_file, calib_key, setup, calib_id, detname \
```
Assigning `raw_files` to `self` and having it constantly update based on the current processing step feels unstable to me.  I suggest we do something different when we actually implement the `state` functionality.

profxj replied:
yes, it is.. a fair bit unstable

kbwestfall commented on pypeit/outputfiles.py:
```diff
@@ -0,0 +1,119 @@
+    basename = fitstbl.construct_basename(frame)
+    ext = '.txt' if txt else '.fits'
+    # Finish``
+    return science_path(par) / f'spec{"2" if twod else "1"}d_{basename}{ext}'
```
Change `science_par(path)` to `sci_path`?  And would it make sense to make `ext` the keyword argument (with a default of `'.fits'`) instead of `txt`?

profxj replied:
good catch;  and yes good call

kbwestfall commented on pypeit/pypeit.py:
```diff
@@ -174,141 +169,53 @@ def qa_path(self) -> str:
 
-                self.caliBrate = self.calib_one(grp_frames, self.det)
-                if not self.caliBrate.success:
+                #self.caliBrate = self.calib_one(grp_frames, self.det, calib_ID)
```
remove?

profxj replied:
gone

kbwestfall commented on pypeit/pypeit.py:
```diff
@@ -174,141 +169,53 @@ def qa_path(self) -> str:
+                if not caliBrate.success:
                     msgs.warn(f'Calibrations for detector {self.det} were unsuccessful!  The step '
-                              f'that failed was {self.caliBrate.failed_step}.  Continuing to next '
+                              f'that failed was {caliBrate.failed_step}.  Continuing to next '
```
Should we put this status check into `calib_one`?  Then `calib_one` wouldn't need to return anything.

profxj replied:
done

kbwestfall commented on pypeit/pypeit.py:
```diff
@@ -1271,16 +291,141 @@ def print_end_time(self):
         return '<{:s}: pypeit_file={}>'.format(self.__class__.__name__, self.pypeit_file)
 
 
+def reduce_calibID(spectrograph, par, fitstbl, calib_ID:str, 
```
Should this be in `pypeit_steps.py`?

profxj replied:
I think it belongs better here

kbwestfall commented on pypeit/pypeit_steps.py:
```diff
@@ -0,0 +1,908 @@
+
+def instantiate_objfind(sciImg, spectrograph, fitstbl, par, frames, det,
+                        caliBrate, bkg_redux, find_negative, 
+                        show:bool=False):
```
It feels odd that we have a function that instantiates the `FindObjects` class.  Can we consolidate this with the `__init__` function of the class?

profxj replied:
not sure I know how..

kbwestfall reviewed (changes requested):
I have a few recommendations, but none of them are show stoppers.  I think this is a step in the right direction, and I'm happy with this as long as the tests pass.  I didn't pay too much attention to "new" code if I expected the changes were largely just based on the file location (e.g., much of what's in `pypeit_steps.py` and `exposure.py`).

I think my biggest big-picture issue (which we should likely leave for a future PR) is that the `pypeit_steps.py` and `exposure.py` feel like grab bags of functions that are not tied together by the type of processing being performed but instead by the level at which they happen during processing.  As a user, I think I would want functions grouped more by type.

kbwestfall committed:
minor

kbwestfall committed:
test fix

profxj committed:
working through the PR

profxj committed:
lots more PR changes

profxj committed:
wip

debora-pe commented on pypeit/core/findobj_skymask.py:
```diff
@@ -1413,8 +1416,11 @@ def objfind_QA(spat_peaks, snr_peaks, spat_vector, snr_vector, snr_thresh, qa_ti
     if objfindQA_filename is not None:
-        fig.savefig(objfindQA_filename, dpi=400)
+        qafile = Path(objfindQA_filename).absolute()
+        qafile.parent.mkdir(parents=True, exists_ok=True)
```
```suggestion
        qafile.parent.mkdir(parents=True, exist_ok=True)
```

badpandabear replied:
I committed this change to get the dev-suite to run

debora-pe replied:
Thanks!

debora-pe commented on pypeit/spectrographs/spectrograph.py:
```diff
@@ -1100,12 +1103,32 @@ def select_detectors(self, subset=None):
+            [1, 2, 3, 4, 5, 6, 7, 8]
+            >>> spectrograph.select_detectors(subset=3)
+            [3]
+            >>> spectrograph.select_detectors(subset=(3,4))
```
Should the result of this be: `[(3,4)]` as per the code on line 1152?
```
        elif isinstance(subset, (int, tuple)):
            _subset = [subset]
```

profxj replied:
I think it can be a tuple, and if it is it is converted to a list

kbwestfall replied:
Here's what I get on this branch:
```
>>> spec.select_detectors(subset=(3,4))
[WARNING] :: Selected detectors or detector mosaics contain invalid values.

[ERROR]   :: The allowed values are det=[1, 2, 3, 4, 5, 6, 7, 8, (1, 5), (2, 6), (3, 7), (4, 8)]
Traceback (most recent call last):
  File "<python-input-3>", line 1, in <module>
    spec.select_detectors(subset=(3,4))
    ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^
  File "/Users/westfall/Work/packages/pypeit-main/pypeit/pypeit/spectrographs/spectrograph.py", line 1178, in select_detectors
    msgs.error(f"The allowed values are det={allowed}")
    ~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/westfall/Work/packages/pypeit-main/pypeit/pypeit/pypmsgs.py", line 195, in error
    raise eval(cls)(msg)
pypeit.pypmsgs.PypeItError: The allowed values are det=[1, 2, 3, 4, 5, 6, 7, 8, (1, 5), (2, 6), (3, 7), (4, 8)]
```
You get the same thing if you run `spec.select_detectors(subset=[(3,4)])` instead, and I think this is the right behavior.  I.e., tuples of detectors are used to indicate mosaics, and a `(3,4)` mosaic is not currently allowed for DEIMOS.

FWIW,
```
>>> spec.select_detectors(subset=(3,7))
[(3, 7)]
>>> spec.select_detectors(subset=[(3,7)])
[(3, 7)]
```
both work fine.

profxj committed:
typo

debora-pe commented on pypeit/metadata.py:
```diff
@@ -619,6 +563,37 @@ def get_configuration_names(self, ignore=None, return_index=False, configs=None)
 
         return (setups, indx) if return_index else setups
 
+    def get_frames_from_combid(self, comb_id, exclude_bkg=False):
```
the parameter `exclude_bkg` is not used?

profxj replied:
gone

debora-pe commented on pypeit/pypeit.py:
```diff
@@ -5,42 +5,29 @@
+
 from pypeit import inputfiles
-from pypeit.calibframe import CalibFrame
+from pypeit import outputfiles
```
This is imported twice, see line 27.

profxj replied:
agreed; gone

debora-pe commented on pypeit/pypeit_steps.py:
```diff
@@ -0,0 +1,914 @@
+            List of indices corresponding to the background frames in the
+            `fitstbl`. If None or empty, no A-B background subtraction is performed.
+            Default is None.
+        sci_outfile (:obj:`str`, optional):
```
This should be a `pathlib.Path` object (not `str`)

profxj replied:
I'm sorry, I can't seem to learn `Path` objects

debora-pe commented on pypeit/pypeit_steps.py:
```diff
@@ -0,0 +1,914 @@
+    return caliBrate
+
+
+def load_skyregions(initial_slits=False, scifile=None, frame=None):
```
This method needs some adjustment. As you copied this from `pypeit.py` there are now some variables that do not exist, like `self.par`, `self.fitstbl`, `self.objtype`.

profxj replied:
gosh, that was a mess

fixed (hopefully)

debora-pe commented on pypeit/exposure.py:
```diff
@@ -0,0 +1,536 @@
+    # Extract
+    for i,det in enumerate(detectors):
+        # Load calibrations
+        caliBrate = pypeit_steps.load_calibrations_for_frame(
```
This `calibrate` is not used.

profxj replied:
good catch; gone

debora-pe commented on pypeit/exposure.py:
```diff
@@ -0,0 +1,536 @@
+        all_specobjs_objfind, calib_slits, spat_flexure, platescale,
+        par['reduce']['slitmask'], par['reduce']['findobj']['find_fwhm'])
+
+    return calib_slits
```
This should return also `all_specobjs_objfind` because those are the detected object + the force-detected object from the slitmask design

profxj replied:
ah, ok;  I thought it was modified in place, but maybe not

debora-pe commented on pypeit/exposure.py:
```diff
@@ -0,0 +1,536 @@
+    # slitmask stuff
+    if par['reduce']['slitmask']['assign_obj']:
+        frame0 = frames[0]
+        calib_slits = adjust_for_slitmask(
```
this should return `calib_slits` and an updated version of `all_specobjs_find`

debora-pe commented on pypeit/scripts/reduce_by_step.py:
```diff
@@ -0,0 +1,248 @@
+                    msgs.error("Need to implement this!")
+
+                # Run me
+                calib_slits = exposure.adjust_for_slitmask(
```
This looks good. The only thing is that this should return also `specobjs_objfind`

debora-pe reviewed (changes requested):
@profxj  here are my comments. I am also running some reductions to make some comparison with previous versions.

debora-pe commented on pypeit/pypeit_steps.py:
```diff
@@ -0,0 +1,914 @@
+            = get_sci_metadata(spectrograph, fitstbl, frames[0], det)
+
+    # Grab the calibrations
+    caliBrate = load_calibrations_for_frame(
```
@profxj I found a problem here, while I was testing. Since the calibrations are loaded from file here, the bpm slit flagging happening during obj finding is not propagated here. This causes inconsistency between the information form `slits` and what is saved in `specobj` from objfind and cause the code to crash. I wonder if we need to save to file also an intermediate slittrace object? This problem I think will happen also if we want to use the maskdesign information in a step-by-step reduction.

debora-pe replied:
Or I don't know if I am getting confused here. But for sure, the masking from `objfind` is not properly propagated to `extract`.

profxj replied:
ok, pinging @kbwestfall too

Let's engage on Slack (on your Monday)

tbowers7 committed:
Replace all instances of ``np.invert`` with ``np.logical_not``

	modified:   pypeit/bspline/util.py
	modified:   pypeit/core/arc.py
	modified:   pypeit/core/coadd.py
	modified:   pypeit/core/extract.py
	modified:   pypeit/core/findobj_skymask.py
	modified:   pypeit/core/fitting.py
	modified:   pypeit/core/flat.py
	modified:   pypeit/core/flexure.py
	modified:   pypeit/core/flux_calib.py
	modified:   pypeit/core/moment.py
	modified:   pypeit/core/skysub.py
	modified:   pypeit/core/trace.py
	modified:   pypeit/core/tracewave.py
	modified:   pypeit/find_objects.py
	modified:   pypeit/flatfield.py
	modified:   pypeit/sensfunc.py
	modified:   pypeit/slittrace.py
	modified:   pypeit/spec2dobj.py
	modified:   pypeit/spectrographs/spectrograph.py
	modified:   pypeit/tracepca.py
	modified:   pypeit/wavecalib.py
	modified:   pypeit/wavetilts.py

Commit:
Fixes so that the dev-suite can test the refactored PypeIt

profxj committed:
Merge remote-tracking branch 'origin/develop' into refactor_pypeit

profxj committed:
getting there

kbwestfall committed:
Merge pull request #2032 from tbowers7/np_invert

Replace all instances of ``np.invert`` with ``np.logical_not``

profxj committed:
Merge branch 'refactor_pypeit' of https://github.com/pypeit/PypeIt into refactor_pypeit

Commit:
Fix import errors

Commit:
Bug fixes based on dev-suite testing

profxj committed:
wip

profxj committed:
puzzled..

debora-pe committed:
fixed one about maskdesign

debora-pe committed:
small fix

debora-pe committed:
Merge branch 'develop' into refactor_pypeit

debora-pe committed:
small merge fix

debora-pe committed:
fix it?

debora-pe committed:
small fix

debora-pe committed:
deal with reduce_by_step and more fixes

debora-pe committed:
cleanup+inline_comment

profxj committed:
skysub doc

kbwestfall committed:
Merge branch 'develop' into refactor_pypeit

kbwestfall committed:
doc update/fixes

debora-pe committed:
fix ifu test failure

debora-pe committed:
solution for IFU?

debora-pe committed:
fix a small one

debora-pe committed:
change how to propagate slitshift

debora-pe committed:
small fix

debora-pe committed:
rjc comments

debora-pe committed:
Merge branch 'develop' into refactor_pypeit_dp

debora-pe committed:
Merge branch 'refactor_pypeit' into refactor_pypeit_dp

debora-pe committed:
Merge pull request #2039 from pypeit/refactor_pypeit_dp

Edits to Refactor PypeIt

debora-pe committed:
Merge branch 'develop' into refactor_pypeit

kbwestfall reviewed (approved):
Approving!

debora-pe commented:
OK. Merging. It's prudent to re-run the tests on the `develop` branch. We'll do that.
